neuro_san/internals/tool_factories/service_tool_factory_provider.py

```diff
@@ -46,6 +48,19 @@ def get_instance(cls):
             ServiceToolFactoryProvider.instance = ServiceToolFactoryProvider()
         return ServiceToolFactoryProvider.instance
 
+    def add_listener(self, listener: AgentStateListener):
+        """
+        Add a state listener to be notified when status of service agents changes.
+        """
+        self.listeners.append(listener)
+
+    def remove_listener(self, listener: AgentStateListener):
+        """
+        Remove a state listener from registered set.
+        """
+        if listener in self.listeners:
+            self.listeners.remove(listener)
+
     def build_agent_tool_registry(self, agent_name: str, config: Dict[str, Any]):
```

Comment:
When/where does build_agent_tool_registry() get called now?
I think it might be removeable.